src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs

```diff
@@ -184,16 +184,25 @@ private void MethodChecks(MethodDeclarationSyntax syntax, Binder withTypeParamsB
             var location = this.Locations[0];
             if (IsAsync)
             {
-                // Warn for CancellationToken parameters in async-iterators with no parameter decorated with [EnumeratorCancellation]
                 var cancellationTokenType = DeclaringCompilation.GetWellKnownType(WellKnownType.System_Threading_CancellationToken);
                 var iAsyncEnumerableType = DeclaringCompilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerable_T);
+                var enumeratorCancellationCount = Parameters.Count(p => p is SourceComplexParameterSymbol { HasEnumeratorCancellationAttribute: true });
                 if (ReturnType.OriginalDefinition.Equals(iAsyncEnumerableType) &&
-                    (Bodies.blockBody != null || Bodies.arrowBody != null) &&
-                    ParameterTypesWithAnnotations.Any(p => p.Type.Equals(cancellationTokenType)) &&
-                    !Parameters.Any(p => p is SourceComplexParameterSymbol { HasEnumeratorCancellationAttribute: true }))
+                    (Bodies.blockBody != null || Bodies.arrowBody != null))
                 {
-                    // There could be more than one parameter that could be decorated with [EnumeratorCancellation] so we warn on the method instead
-                    diagnostics.Add(ErrorCode.WRN_UndecoratedCancellationTokenParameter, location, this);
+                    if (ParameterTypesWithAnnotations.Any(p => p.Type.Equals(cancellationTokenType)) &&
+                        enumeratorCancellationCount == 0)
```

Comment:
I would swap the order of these conditions; the latter is very cheap and the former more expensive.  Also, generally speaking we prefer to avoid Linq methods (e.g. to avoid the delegate creation here)